pkg/proxy/healthcheck/service_health.go

```diff
@@ -117,25 +141,75 @@ func (hcs *server) SyncServices(newServices map[types.NamespacedName]uint16) err
 			continue
 		}
 		hcs.services[nsn] = svc
+	}
+	return nil
+}
+
+type hcInstance struct {
+	nsn  types.NamespacedName
+	port uint16
+
+	listeners   []net.Listener
+	httpServers []httpServer
 
-		go func(nsn types.NamespacedName, svc *hcInstance) {
+	endpoints int // number of local endpoints for a service
+}
+
+// listenAll opens health check port on all the addresses provided
+func (hcI *hcInstance) listenAndServeAll(hcs *server) error {
+	var err error
+	var listener net.Listener
+
+	addresses := hcs.nodeAddresses.List()
+	hcI.listeners = make([]net.Listener, 0, len(addresses))
+	hcI.httpServers = make([]httpServer, 0, len(addresses))
+
+	// for each of the node addresses start listening and serving
+	for _, address := range addresses {
+		addr := fmt.Sprintf("%s:%d", address, hcI.port)
+		// create http server
+		httpSrv := hcs.httpFactory.New(addr, hcHandler{name: hcI.nsn, hcs: hcs})
+		// start listener
+		listener, err = hcs.listener.Listen(addr)
+		if err != nil {
+			// must close whatever have been previously opened
+			// to allow a retry/or port ownership change as needed
+			_ = hcI.closeAll()
+			return err
+		}
+
+		// start serving
+		go func(hcI *hcInstance, listener net.Listener, httpSrv httpServer) {
 			// Serve() will exit when the listener is closed.
-			klog.V(3).Infof("Starting goroutine for healthcheck %q on port %d", nsn.String(), svc.port)
-			if err := svc.server.Serve(svc.listener); err != nil {
-				klog.V(3).Infof("Healthcheck %q closed: %v", nsn.String(), err)
+			klog.V(3).Infof("Starting goroutine for healthcheck %q on  %s", hcI.nsn.String(), listener.Addr().String())
+			if err := httpSrv.Serve(listener); err != nil {
+				klog.V(3).Infof("Healthcheck %q closed: %v", hcI.nsn.String(), err)
 				return
 			}
-			klog.V(3).Infof("Healthcheck %q closed", nsn.String())
-		}(nsn, svc)
+			klog.V(3).Infof("Healthcheck %q on %s closed", hcI.nsn.String(), listener.Addr().String())
+		}(hcI, listener, httpSrv)
+
+		hcI.listeners = append(hcI.listeners, listener)
+		hcI.httpServers = append(hcI.httpServers, httpSrv)
```

Comment:
if the httpSrv contains the listener, do we need to store both? 

can it be racy if we close a server and a listener for that server?

Reply:
nvm, we only close the listeners